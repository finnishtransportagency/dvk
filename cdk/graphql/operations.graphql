query findFairways {
  fairways {
    ...fragment FairwayFragment on Fairway {
      id
      nameFI
      nameSV
    }
  }
}

query findFairway($id: Int) {
  fairway(id: $id) {
    ...Fairway
  }
}
