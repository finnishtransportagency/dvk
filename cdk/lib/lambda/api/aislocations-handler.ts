import { ALBEvent, ALBResult } from 'aws-lambda';
import { log } from '../logger';
import { fetchLocations } from './ais';
import { handleAisCall } from '../util';

export const handler = async (event: ALBEvent): Promise<ALBResult> => {
  log.info({ event }, `aislocations()`);
  return handleAisCall('aislocations', fetchLocations, ['application/geo+json']);
};
