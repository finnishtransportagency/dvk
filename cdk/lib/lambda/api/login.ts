import { getCloudFrontPrivateKey, getCloudFrontPublicKeyId, getCognitoUrl } from '../environment';
import { getSignedCookies } from '@aws-sdk/cloudfront-signer';
import { ALBEvent, ALBResult } from 'aws-lambda';
import { log } from '../logger';
import JWT, { JwtPayload } from 'jsonwebtoken';
import jwkToPem from 'jwk-to-pem';
import axios from 'axios';

const cloudFrontDnsName = process.env.CLOUDFRONT_DNSNAME;
let cachedKeys: Record<string, string>;

// Fetch JWK's from Cognito or cache
const getPublicKeys = async (issuerUrl: string) => {
  if (!cachedKeys) {
    cachedKeys = {};
    const publicKeys = await axios.get(issuerUrl + '/.well-known/jwks.json');
    for (const key of publicKeys.data.keys) {
      cachedKeys[key.kid] = jwkToPem(key);
    }
    return cachedKeys;
  } else {
    return cachedKeys;
  }
};

export const validateJwtToken = async (token: string | undefined, dataToken: string): Promise<JwtPayload | undefined> => {
  if (!token) {
    log.debug('IAM JWT Token missing');
    return;
  }
  // Split token into parts
  const tokenParts = token.split('.');
  if (tokenParts.length < 2) {
    log.error('Invalid token');
    return;
  }

  // Parse header & payload from token parts
  const tokenHeader = JSON.parse(Buffer.from(tokenParts[0], 'base64').toString('utf-8'));
  const tokenPayload = JSON.parse(Buffer.from(tokenParts[1], 'base64').toString('utf-8'));

  // Fetch public keys from Cognito
  const publicKeys = await getPublicKeys(tokenPayload.iss);
  const publicKey = publicKeys[tokenHeader.kid];
  if (!publicKey) {
    log.debug('Public key not found');
    return;
  }

  // Verify token
  const result = JWT.verify(token, publicKey, { issuer: await getCognitoUrl() }) as JwtPayload;
  if (!result) {
    log.error('Failed to verify JWT');
    return;
  }

  // Check use access
  if (result.token_use !== 'access') {
    log.error('Invalid token use');
    return;
  }

  // Return decoded data token
  return JWT.decode(dataToken.replace(/=/g, '')) as JwtPayload;
};

export const handler = async (event: ALBEvent): Promise<ALBResult> => {
  try {
    let jwtToken;
    if (event.multiValueHeaders) {
      log.info({ headers: event.multiValueHeaders }, 'Request headers');
      const token = event.multiValueHeaders['x-iam-accesstoken'];
      const data = event.multiValueHeaders['x-iam-data'];
      if (token && data) {
        jwtToken = await validateJwtToken(token[0], data[0]);
      }
    }
    log.info('JwtToken: %o', jwtToken);
    // TODO: comment out once token received from ALB
    /*if (!jwtToken) {
      return {
        statusCode: 403,
      };
    }*/
    const cloudFrontPolicy = JSON.stringify({
      Statement: [
        {
          Resource: `https://${cloudFrontDnsName}/*`,
          Condition: {
            DateLessThan: {
              'AWS:EpochTime': Math.floor(new Date().getTime() / 1000) + 3600,
            },
          },
        },
      ],
    });

    const cookie = getSignedCookies({
      url: cloudFrontDnsName as string,
      privateKey: await getCloudFrontPrivateKey(),
      keyPairId: await getCloudFrontPublicKeyId(),
      policy: cloudFrontPolicy,
    });

    const setCookieAttributes = `; Domain=${cloudFrontDnsName}; Path=/; Secure; HttpOnly; SameSite=Lax`;

    const response: ALBResult = {
      statusCode: 302,
      multiValueHeaders: {
        Location: [`https://${cloudFrontDnsName}/yllapito/index.html`],
        'Set-Cookie': [
          `CloudFront-Key-Pair-Id=${cookie['CloudFront-Key-Pair-Id']}${setCookieAttributes}`,
          `CloudFront-Policy=${cookie['CloudFront-Policy']}${setCookieAttributes}`,
          `CloudFront-Signature=${cookie['CloudFront-Signature']}${setCookieAttributes}`,
        ],
      },
    };
    log.debug(response, 'Login response');
    return response;
  } catch (e) {
    log.error('Getting signed cookies failed: %s', e);
    return { statusCode: 500, body: String(e) };
  }
};
