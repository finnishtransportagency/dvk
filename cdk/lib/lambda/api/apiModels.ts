import { Geometry } from 'geojson';
import { GeometryPoint, Text } from '../../../graphql/generated';
import { FairwayCardIdName } from '../db/fairwayCardDBModel';

export type VaylaGeojsonFeature = { type?: string; geometry: Geometry };

/* VATU */
type MitoitusAlusAPIModel = {
  alustyyppiKoodi: string;
  alustyyppi: string;
  pituus: number;
  leveys: number;
  syvays: number;
  koko?: number;
  runkoTaytelaisyysKerroin?: number;
};

type LuokitusAPIModel = {
  luokitusTyyppi: string;
  vaylaluokkaKoodi: string;
  vaylaluokkaFI?: string;
  vaylaluokkaSV?: string;
};

export type VaylaAPIModel = {
  jnro: number;
  nimiFI: string;
  nimiSV?: string;
  vaylalajiKoodi?: string;
  vaylaLajiFI?: string;
  vaylaLajiSV?: string;
  valaistusKoodi?: string;
  valaistusFI?: string;
  valaistusSV?: string;
  omistaja?: string;
  merialueFI?: string;
  merialueSV?: string;
  alunSeloste?: string;
  paatepisteenSeloste?: string;
  normaaliKaantosade?: number;
  minimiKaantosade?: number;
  normaaliLeveys?: number;
  minimiLeveys?: number;
  varavesi?: string;
  lisatieto?: string;
  mareografi?: string;
  mitoitusalus?: MitoitusAlusAPIModel[];
  luokitus?: LuokitusAPIModel[];
};

export type VaylaFeature = { properties: VaylaAPIModel };
export type VaylaFeatureCollection = { type?: string; features: VaylaFeature[] };

type AlueVaylaAPIModel = {
  jnro: number;
  status: number;
  linjaus: number;
  mitoitusNopeus?: number;
  mitoitusNopeus2?: number;
  vaylaalueenJarjestysNro?: number;
  nimiFI?: string;
  nimiSV?: string;
};

type AlueProperties = {
  id: number;
  nimi?: string;
  mitoitusSyvays?: number;
  harausSyvyys?: number;
  vertaustaso?: string;
  n2000MitoitusSyvays?: number;
  n2000HarausSyvyys?: number;
  n2000Vertaustaso?: string;
  suunta?: number;
  diaariNumero?: string;
  vahvistusPaivamaara?: string;
  omistaja?: string;
  lisatieto?: string;
  vayla?: AlueVaylaAPIModel[];
  tyyppiKoodi?: number;
  tyyppi?: string;
  merkintalajiKoodi?: number;
  merkintalaji?: string;
  liikennointiStatusKoodi?: number;
  liikennointiStatus?: string;
  liikennointiTyyppiKoodi?: number;
  liikenteenTyyppi?: string;
  liikennointiSuuntaKoodi?: number;
  liikennointiSuunta?: string;
};

export type AlueFeature = {
  properties: AlueProperties;
} & VaylaGeojsonFeature;

export type AlueFeatureCollection = {
  type?: string;
  features: AlueFeature[];
};

type NavigointiLinjaVaylaAPIModel = {
  jnro: number;
  status?: number;
  linjaus?: number;
  nimiFI?: string;
  nimiSV?: string;
};

type NavigointiLinjaProperties = {
  id: number;
  mitoitusSyvays?: number;
  harausSyvyys?: number;
  vertaustaso?: string;
  n2000MitoitusSyvays?: number;
  n2000HarausSyvyys?: number;
  n2000Vertaustaso?: string;
  tosisuunta?: number;
  pituus?: number;
  diaariNumero?: string;
  vahvistusPaivamaara?: string;
  omistaja?: string;
  lisatieto?: string;
  tyyppiKoodi?: string;
  tyyppi?: string;
  vayla: NavigointiLinjaVaylaAPIModel[];
};

export type NavigointiLinjaFeature = {
  properties: NavigointiLinjaProperties;
} & VaylaGeojsonFeature;

export type NavigointiLinjaFeatureCollection = {
  type?: string;
  features: NavigointiLinjaFeature[];
};

type RajoitustyyppiAPIModel = {
  koodi?: string;
  rajoitustyyppi?: string;
};

type RajoitusVaylaAPIModel = {
  jnro: number;
  nimiFI?: string;
  nimiSV?: string;
};

type RajoitusAlueProperties = {
  id: number;
  rajoitustyyppi?: string;
  rajoitustyypit?: RajoitustyyppiAPIModel[];
  suuruus?: number;
  esittaja?: string;
  diaariNumero?: string;
  vahvistusPaivamaara?: string;
  muutosPaivamaara?: string;
  alkuPaivamaara?: string;
  loppuPaivamaara?: string;
  paatosTila?: string;
  tietolahde?: string;
  sijainti?: string;
  kunta?: string;
  poikkeus?: string;
  vayla?: RajoitusVaylaAPIModel[];
};

export type RajoitusAlueFeature = { properties: RajoitusAlueProperties } & VaylaGeojsonFeature;

export type RajoitusAlueFeatureCollection = {
  type?: string;
  features: RajoitusAlueFeature[];
};

export type TurvalaiteVaylaAPIModel = {
  jnro: number;
  paavayla?: string;
};

export type TurvalaiteVikatiedotProperties = {
  vikaId: number;
  turvalaiteNumero: number;
  turvalaiteNimiFI: string;
  turvalaiteNimiSV?: string;
  vikatyyppiKoodi: number;
  vikatyyppiFI: string;
  vikatyyppiSV?: string;
  vikatyyppiEN?: string;
  kirjausAika: string;
};
export type TurvalaiteVikatiedotFeature = { properties: TurvalaiteVikatiedotProperties } & VaylaGeojsonFeature;
export type TurvalaiteVikatiedotFeatureCollection = {
  type?: string;
  features: TurvalaiteVikatiedotFeature[];
};

type TurvalaiteReunaetaisyysAPIModel = {
  vaylaalueID: number;
  etaisyys: number;
};

type TurvalaiteProperties = {
  turvalaitenumero: number;
  nimiFI?: string;
  nimiSV?: string;
  alityyppi?: string;
  turvalaitetyyppiKoodi?: number;
  turvalaitetyyppiFI?: string;
  turvalaitetyyppiSV?: string;
  navigointilajiKoodi?: number;
  navigointilajiFI?: string;
  navigointilajiSV?: string;
  valaistu?: string;
  omistajaFI?: string;
  omistajaSV?: string;
  symboli?: string;
  kaukohallinta?: number;
  AISTyyppi?: number;
  AISTyyppiSeliteFI?: string;
  AISTyyppiSeliteSV?: string;
  vayla?: TurvalaiteVaylaAPIModel[];
  reunaetaisyys?: TurvalaiteReunaetaisyysAPIModel[];
};

export type TurvalaiteFeature = { properties: TurvalaiteProperties } & VaylaGeojsonFeature;

export type TurvalaiteFeatureCollection = {
  type?: string;
  features: TurvalaiteFeature[];
};

export type TaululinjaVaylaAPIModel = {
  jnro: number;
  nimiFI?: string;
  nimiSV?: string;
};

export type TaululinjaProperties = {
  taululinjaId: number;
  suunta: number;
  etuTLId?: number;
  keskiTLId?: number;
  takaTLId?: number;
  vayla: TaululinjaVaylaAPIModel[];
};

export type TaululinjaFeature = { properties: TaululinjaProperties } & VaylaGeojsonFeature;
export type TaululinjaFeatureCollection = {
  type?: string;
  features: TaululinjaFeature[];
};

type KaantoympyraVaylaAPIModel = {
  jnro: number;
  nimiFI: string;
  nimiSV?: string;
  luokitus: number;
};

export type KaantoympyraProperties = {
  kaantoympyraID: number;
  halkaisija: number;
  haraussyvyys?: number;
  vertaustaso?: string;
  lisatieto?: string;
  vayla?: KaantoympyraVaylaAPIModel[];
};

export type KaantoympyraFeature = { properties: KaantoympyraProperties } & VaylaGeojsonFeature;

export type KaantoympyraFeatureCollection = { type?: string; features: KaantoympyraFeature[] };

/* Traficom */
export type PilotPlace = {
  id: number;
  name: Text;
  geometry: GeometryPoint;
  fairwayCards: FairwayCardIdName[];
};

/* Pooki */
export type MarineWarningDates = {
  startDateTime?: number;
  endDateTime?: number;
  dateTime?: number;
};

/* Weather */
export type WeatherMareograph = {
  fmisid: number;
  geoid: number;
  latlon: string;
  station_name: string;
  localtime: string;
  WLEV_PT1S_INSTANT: number;
  WLEVN2K_PT1S_INSTANT: number;
};

export type Mareograph = {
  id: number;
  geometry: Geometry;
  name: string;
  dateTime: number;
  waterLevel: number;
  n2000WaterLevel: number;
  calculated: boolean;
};

export type WeatherObservation = {
  fmisid: number;
  geoid: number;
  latlon: string;
  stationname: string;
  localtime: string;
  WG_PT10M_MAX: number;
  WD_PT10M_AVG: number;
  WS_PT10M_AVG: number;
  TA_PT1M_AVG: number;
  VIS_PT1M_AVG: number | null;
};

export type Observation = {
  id: number;
  name: string;
  windSpeedAvg: number;
  windSpeedMax: number;
  windDirection: number;
  visibility: number | null;
  temperature: number;
  dateTime: number;
  geometry: Geometry;
};

export type WeatherBuoy = {
  fmisid: number;
  geoid: number;
  latlon: string;
  station_name: string;
  localtime: string;
  WH_PT1M_ACC: number | null;
  WHD_PT1M_ACC: number | null;
  TW_PT1M_AVG: number | null;
};

export type Buoy = {
  id: number;
  name: string;
  dateTime: number;
  geometry: Geometry;
  temperature: number | null;
  waveDirection: number | null;
  waveHeight: number | null;
};

/* AIS */
export type VesselAPIModel = {
  name: string;
  timestamp: number;
  mmsi: number;
  callSign: string;
  imo: number;
  shipType: number;
  draught: number;
  eta: number;
  posType: number;
  referencePointA: number;
  referencePointB: number;
  referencePointC: number;
  referencePointD: number;
  destination: string;
};

type VesselLocationProperties = {
  mmsi: number;
  sog: number;
  cog: number;
  navStat: number;
  rot: number;
  posAcc: boolean;
  raim: boolean;
  heading?: number;
  timestamp: number;
  timestampExternal: number;
};

type VesselLocationFeature = {
  mmsi: number;
  type?: string;
  geometry: Geometry;
  properties: VesselLocationProperties;
};

export type VesselLocationFeatureCollection = {
  type?: string;
  dataUpdatedTime?: string;
  features: VesselLocationFeature[];
};

export type Vessel = {
  name: string;
  timestamp: Date;
  mmsi: number;
  callSign: string;
  imo: number;
  shipType: number;
  draught: number; // meters
  eta: string; // MM-DD HH:mm
  posType: number;
  referencePointA: number;
  referencePointB: number;
  referencePointC: number;
  referencePointD: number;
  destination: string;
};

/* RTZ */
export type Coordinate = [number, number];

export type RtzWaypoint = {
  coordinate: Coordinate;
  turnRadius: number;
};

type RtzGeometria = {
  type: string;
  coordinates: Coordinate;
};

export type RtzReittipiste = {
  tunnus: number;
  nimi: string;
  rtzTunniste: number;
  reittitunnus: number;
  kaarresade: number;
  geometria: RtzGeometria;
  leveysVasen: number;
  leveysOikea: number;
  geometriaTyyppi: string;
  muutosaikaleima: string;
  jarjestys: number;
};

export type RtzData = {
  tunnus: number;
  tila: number;
  nimi: string;
  tunniste: string;
  rtz: string;
  reittipisteet: Array<RtzReittipiste>;
};

/* IBNet */
export type IBNetApiResponse = {
  rvEndpoints: string[];
  toRv: string;
};

export type Dirway = {
  id: string;
  name: string;
  description: string;
  rv: number | null;
  change_time: string;
  deleted?: boolean | null;
};

export type DirwayPoint = {
  id: string;
  dirway_id: string;
  order_num: number;
  name: string;
  latitude: number;
  longitude: number;
  rv: number | null;
  change_time: string;
  deleted?: boolean | null;
};

export type LocationApiModel = {
  id: string;
  name: string;
  type: 'PORT' | 'FAIRWAY';
  locode_list: string;
  nationality: string;
  latitude: number;
  longitude: number;
  winterport: boolean;
  rv: number | null;
  change_time: string;
  deleted?: boolean | null;
};

export type RestrictionApiModel = {
  id: string;
  location_id: string;
  start_time: string;
  end_time?: string | null;
  text_compilation: string;
  rv: number | null;
  change_time: string;
  deleted?: boolean | null;
};
