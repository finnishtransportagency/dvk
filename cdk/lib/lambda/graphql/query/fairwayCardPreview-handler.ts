import { AppSyncResolverEvent, AppSyncResolverHandler } from 'aws-lambda/trigger/appsync-resolver';
import { FairwayCard, QueryFairwayCardPreviewArgs, Status } from '../../../../graphql/generated';
import { log } from '../../logger';
import FairwayCardDBModel from '../../db/fairwayCardDBModel';
import { getPilotPlaceMap, mapFairwayCardDBModelToGraphqlType } from '../../db/modelMapper';
import { ADMIN_ROLE, getOptionalCurrentUser } from '../../api/login';

export const handler: AppSyncResolverHandler<QueryFairwayCardPreviewArgs, FairwayCard | undefined> = async (
  event: AppSyncResolverEvent<QueryFairwayCardPreviewArgs>
): Promise<FairwayCard | undefined> => {
  log.info(`fairwayCard(${event.arguments.id})`);
  const user = await getOptionalCurrentUser(event);

  if (user?.roles.includes(ADMIN_ROLE)) {
    const pilotMap = await getPilotPlaceMap();
    const dbModel = await FairwayCardDBModel.get(event.arguments.id);

    if (dbModel?.status === Status.Draft || dbModel?.status === Status.Public) {
      return mapFairwayCardDBModelToGraphqlType(dbModel, pilotMap, user);
    }
  }

  return undefined;
};
