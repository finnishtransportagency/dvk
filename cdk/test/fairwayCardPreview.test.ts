import { mockClient } from 'aws-sdk-client-mock';
import { DynamoDBDocumentClient, GetCommand } from '@aws-sdk/lib-dynamodb';
import { handler as previewHandler } from '../lib/lambda/graphql/query/fairwayCardPreview-handler';
import { Status } from '../graphql/generated';
import { mockContext, mockQueryByIdEvent } from './mocks';
import FairwayCardDBModel from '../lib/lambda/db/fairwayCardDBModel';
import { pilotPlaceMap } from '../lib/lambda/db/modelMapper';
import { ADMIN_ROLE, getOptionalCurrentUser } from '../lib/lambda/api/login';
import { RtzData } from '../lib/lambda/api/apiModels';
import { FeatureCollection } from 'geojson';

const ddbMock = mockClient(DynamoDBDocumentClient);

const card1: FairwayCardDBModel = {
  id: 'public',
  version: 'v0',
  name: {
    fi: 'Testfi',
    sv: 'Testsv',
    en: 'Testen',
  },
  creator: 'test',
  creationTimestamp: Date.now(),
  modifier: 'test2',
  modificationTimestamp: Date.now(),
  status: Status.Public,
  fairways: [{ id: 1, primary: true, primarySequenceNumber: 1, secondary: false, secondarySequenceNumber: 1 }],
  trafficService: {
    pilot: {
      places: [{ id: 1, pilotJourney: 1 }],
    },
  },
  pilotRoutes: [{ id: 1 }],
};

const card2: FairwayCardDBModel = {
  id: 'draft',
  version: 'v0',
  name: {
    fi: 'Testfi2',
    sv: 'Testsv2',
    en: 'Testen2',
  },
  creator: 'test2',
  creationTimestamp: Date.now(),
  modifier: 'test',
  modificationTimestamp: Date.now(),
  status: Status.Draft,
  fairways: [{ id: 1, primary: true, secondary: false }],
  trafficService: {
    pilot: {
      places: [{ id: 2, pilotJourney: 2 }],
    },
  },
  pilotRoutes: [{ id: 2 }],
};

const card3: FairwayCardDBModel = {
  id: 'removed',
  version: 'v0',
  name: {
    fi: 'Testfi3',
    sv: 'Testsv3',
    en: 'Testen3',
  },
  creator: 'test3',
  creationTimestamp: Date.now(),
  modifier: 'test',
  modificationTimestamp: Date.now(),
  status: Status.Removed,
  fairways: [{ id: 1, primary: true, secondary: false }],
  trafficService: {
    pilot: {
      places: [{ id: 3, pilotJourney: 3 }],
    },
  },
  pilotRoutes: [{ id: 3 }],
};

const points: FeatureCollection = {
  type: 'FeatureCollection',
  features: [
    {
      type: 'Feature',
      geometry: { type: 'Point', coordinates: [26.43, 60.28] },
      properties: { IDENTIFIER: 'FI 1', OBJNAM: 'Luotsipaikka 1' },
    },
    {
      type: 'Feature',
      geometry: { type: 'Point', coordinates: [25.001, 60.21] },
      id: 'FI 2',
      properties: { IDENTIFIER: 'FI 2', OBJNAM: 'Luotsipaikka 2' },
    },
    {
      type: 'Feature',
      geometry: { type: 'Point', coordinates: [25.578, 60.14] },
      id: 'FI 3',
      properties: { IDENTIFIER: 'FI 3', OBJNAM: 'Luotsipaikka 3' },
    },
  ],
};

const routes: RtzData[] = [
  {
    tunnus: 1,
    tila: 1,
    nimi: 'Reitti 1',
    tunniste: 'a',
    rtz: 'xml',
    reittipisteet: [
      {
        tunnus: 10,
        nimi: 'A',
        rtzTunniste: 10,
        reittitunnus: 10,
        kaarresade: 0.1,
        geometria: {
          type: 'Point',
          coordinates: [26.9, 60.4],
        },
        leveysVasen: 0.01,
        leveysOikea: 0.01,
        geometriaTyyppi: 'Loxodrome',
        muutosaikaleima: '2024-01-01T01:01:01.000000+02:00',
        jarjestys: 1,
      },
    ],
  },
  {
    tunnus: 2,
    tila: 1,
    nimi: 'Reitti 2',
    tunniste: 'b',
    rtz: 'xml',
    reittipisteet: [
      {
        tunnus: 20,
        nimi: 'B',
        rtzTunniste: 20,
        reittitunnus: 20,
        kaarresade: 0.2,
        geometria: {
          type: 'Point',
          coordinates: [26.9, 60.4],
        },
        leveysVasen: 0.02,
        leveysOikea: 0.02,
        geometriaTyyppi: 'Loxodrome',
        muutosaikaleima: '2024-02-02T02:02:02.000000+02:00',
        jarjestys: 1,
      },
      {
        tunnus: 21,
        nimi: 'BB',
        rtzTunniste: 21,
        reittitunnus: 21,
        kaarresade: 0.2,
        geometria: {
          type: 'Point',
          coordinates: [26.91, 60.41],
        },
        leveysVasen: 0.02,
        leveysOikea: 0.02,
        geometriaTyyppi: 'Loxodrome',
        muutosaikaleima: '2024-02-02T02:02:02.000000+02:00',
        jarjestys: 2,
      },
    ],
  },
  {
    tunnus: 3,
    tila: 1,
    nimi: 'Reitti 3',
    tunniste: 'c',
    rtz: 'xml',
    reittipisteet: [
      {
        tunnus: 30,
        nimi: 'C',
        rtzTunniste: 30,
        reittitunnus: 30,
        kaarresade: 0.3,
        geometria: {
          type: 'Point',
          coordinates: [26.93, 60.43],
        },
        leveysVasen: 0.03,
        leveysOikea: 0.03,
        geometriaTyyppi: 'Loxodrome',
        muutosaikaleima: '2024-03-03T03:03:03.000000+02:00',
        jarjestys: 1,
      },
      {
        tunnus: 31,
        nimi: 'CC',
        rtzTunniste: 31,
        reittitunnus: 31,
        kaarresade: 0.3,
        geometria: {
          type: 'Point',
          coordinates: [26.93, 60.43],
        },
        leveysVasen: 0.03,
        leveysOikea: 0.03,
        geometriaTyyppi: 'Loxodrome',
        muutosaikaleima: '2024-03-03T03:03:03.000000+02:00',
        jarjestys: 2,
      },
      {
        tunnus: 32,
        nimi: 'CCC',
        rtzTunniste: 32,
        reittitunnus: 32,
        kaarresade: 0.3,
        geometria: {
          type: 'Point',
          coordinates: [26.93, 60.43],
        },
        leveysVasen: 0.03,
        leveysOikea: 0.03,
        geometriaTyyppi: 'Loxodrome',
        muutosaikaleima: '2024-03-03T03:03:03.000000+02:00',
        jarjestys: 3,
      },
    ],
  },
];

const adminUser = {
  uid: 'K123456',
  firstName: 'Developer',
  lastName: 'X',
  roles: [ADMIN_ROLE],
};

const otherUser = {
  uid: 'K654321',
  firstName: 'User',
  lastName: 'X',
  roles: ['DVK_Kayttaja'],
};

jest.mock('../lib/lambda/environment', () => ({
  getEnvironment: () => 'mock',
  isPermanentEnvironment: () => false,
  getFairwayCardTableName: () => 'FairwayCard-mock',
}));

jest.mock('../lib/lambda/api/axios', () => ({
  fetchTraficomApi: () => points,
  fetchPilotRoutesApi: () => routes,
}));

jest.mock('../lib/lambda/graphql/cache', () => ({
  getFromCache: () => {
    return { expired: true };
  },
  cacheResponse: () => Promise.resolve(),
}));

jest.mock('../lib/lambda/api/login');

beforeEach(() => {
  jest.resetAllMocks();
  ddbMock.reset();
  pilotPlaceMap.clear();
  (getOptionalCurrentUser as jest.Mock).mockImplementation(() => adminUser);
});

it('should get public card from the DynamoDB', async () => {
  ddbMock
    .on(GetCommand, {
      Key: { id: 'test' },
    })
    .resolves({
      Item: card1,
    });
  const response = await previewHandler(mockQueryByIdEvent, mockContext, () => {});
  expect(response).toMatchSnapshot({
    modificationTimestamp: expect.any(Number),
    creationTimestamp: expect.any(Number),
  });
});

it('should get draft card from the DynamoDB', async () => {
  ddbMock
    .on(GetCommand, {
      Key: { id: 'test' },
    })
    .resolves({
      Item: card2,
    });
  const response = await previewHandler(mockQueryByIdEvent, mockContext, () => {});
  expect(response).toMatchSnapshot({
    modificationTimestamp: expect.any(Number),
    creationTimestamp: expect.any(Number),
  });
});

it('should filter removed card from the DynamoDB', async () => {
  ddbMock
    .on(GetCommand, {
      Key: { id: 'test' },
    })
    .resolves({
      Item: card3,
    });
  const response = await previewHandler(mockQueryByIdEvent, mockContext, () => {});
  expect(response).toBe(undefined);
});

it('should return nothing when admin role missing', async () => {
  (getOptionalCurrentUser as jest.Mock).mockImplementation(() => otherUser);
  ddbMock
    .on(GetCommand, {
      Key: { id: 'test' },
    })
    .resolves({
      Item: card1,
    });
  const response = await previewHandler(mockQueryByIdEvent, mockContext, () => {});
  expect(response).toBe(undefined);
});
