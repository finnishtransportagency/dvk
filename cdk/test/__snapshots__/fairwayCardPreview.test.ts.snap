// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get draft card from the DynamoDB 1`] = `
{
  "additionalInfo": undefined,
  "anchorage": undefined,
  "attention": undefined,
  "creationTimestamp": Any<Number>,
  "creator": "test2",
  "currentPublic": undefined,
  "designSpeed": undefined,
  "fairwayIds": "#1#",
  "fairways": [
    {
      "id": 1,
      "primary": true,
      "primarySequenceNumber": undefined,
      "secondary": false,
      "secondarySequenceNumber": undefined,
    },
  ],
  "generalInfo": undefined,
  "group": undefined,
  "harbors": undefined,
  "iceCondition": undefined,
  "id": "draft",
  "latest": undefined,
  "latestVersionUsed": undefined,
  "lineText": undefined,
  "mareographs": undefined,
  "modificationTimestamp": Any<Number>,
  "modifier": "test",
  "n2000HeightSystem": false,
  "name": {
    "en": "Testen2",
    "fi": "Testfi2",
    "sv": "Testsv2",
  },
  "navigationCondition": undefined,
  "pictures": undefined,
  "pilotRoutes": [
    {
      "id": 2,
      "name": "Reitti 2",
    },
  ],
  "speedLimit": undefined,
  "status": "DRAFT",
  "temporaryNotifications": [],
  "trafficService": {
    "pilot": {
      "email": undefined,
      "extraInfo": undefined,
      "fax": undefined,
      "internet": undefined,
      "phoneNumber": undefined,
      "places": [
        {
          "geometry": {
            "coordinates": [
              25.001,
              60.21,
            ],
            "type": "Point",
          },
          "id": 2,
          "name": {
            "en": "Luotsipaikka 2",
            "fi": "Luotsipaikka 2",
            "sv": "Luotsipaikka 2",
          },
          "pilotJourney": 2,
        },
      ],
    },
    "tugs": null,
    "vts": null,
  },
  "version": "v0",
  "vesselRecommendation": undefined,
  "visibility": undefined,
  "windRecommendation": undefined,
}
`;

exports[`should get public card from the DynamoDB 1`] = `
{
  "additionalInfo": undefined,
  "anchorage": undefined,
  "attention": undefined,
  "creationTimestamp": Any<Number>,
  "creator": "test",
  "currentPublic": undefined,
  "designSpeed": undefined,
  "fairwayIds": "#1#",
  "fairways": [
    {
      "id": 1,
      "primary": true,
      "primarySequenceNumber": 1,
      "secondary": false,
      "secondarySequenceNumber": 1,
    },
  ],
  "generalInfo": undefined,
  "group": undefined,
  "harbors": undefined,
  "iceCondition": undefined,
  "id": "public",
  "latest": undefined,
  "latestVersionUsed": undefined,
  "lineText": undefined,
  "mareographs": undefined,
  "modificationTimestamp": Any<Number>,
  "modifier": "test2",
  "n2000HeightSystem": false,
  "name": {
    "en": "Testen",
    "fi": "Testfi",
    "sv": "Testsv",
  },
  "navigationCondition": undefined,
  "pictures": undefined,
  "pilotRoutes": [
    {
      "id": 1,
      "name": "Reitti 1",
    },
  ],
  "speedLimit": undefined,
  "status": "PUBLIC",
  "temporaryNotifications": [],
  "trafficService": {
    "pilot": {
      "email": undefined,
      "extraInfo": undefined,
      "fax": undefined,
      "internet": undefined,
      "phoneNumber": undefined,
      "places": [
        {
          "geometry": {
            "coordinates": [
              26.43,
              60.28,
            ],
            "type": "Point",
          },
          "id": 1,
          "name": {
            "en": "Luotsipaikka 1",
            "fi": "Luotsipaikka 1",
            "sv": "Luotsipaikka 1",
          },
          "pilotJourney": 1,
        },
      ],
    },
    "tugs": null,
    "vts": null,
  },
  "version": "v0",
  "vesselRecommendation": undefined,
  "visibility": undefined,
  "windRecommendation": undefined,
}
`;
